src/storage/backplane_control.cpp

```diff
@@ -0,0 +1,612 @@
+    {
+        log<level::ERR>(
+            "Only 8-bit ASCII supported for drive FRU Serial Number field",
+            entry("RESULT=%d", res), entry("REASON=%s", std::strerror(-res)));
+        return "";
+    }
+
+    const char* snPtr = reinterpret_cast<char*>(data + offset + 1);
+    return std::string{snPtr, snLen};
+}
+
+constexpr size_t v1aSNFieldOffset = 5;
+constexpr size_t v1aSNFieldSize = 20;
+
+static std::string getNVMeSerialNumberV1A(i2cDev& dev, const unsigned char* buf)
+{
+    static_assert(i2cDev::i2cBlockSize >= v1aSNFieldOffset + v1aSNFieldSize);
+
+    if (!((buf[0] == 0x02) && (buf[1] == 0x08) && (buf[2] == 0x01)))
+    {
+        return "";
+    }
+
+    const char* snPtr = reinterpret_cast<const char*>(buf);
+    return std::string{snPtr + v1aSNFieldOffset, v1aSNFieldSize};
+}
+
+BackplaneController::BackplaneController(
+    sdbusplus::bus::bus& bus, int i2cBus, int i2cAddr,
+    const BackplaneControllerConfig& config) :
+    BackplaneMCUServer(
+        bus, dbusEscape(std::string(dbus::stormgr::path) + "/backplane/MCU_" +
+                        std::to_string(i2cBus) + "_" + std::to_string(i2cAddr))
+                 .c_str()),
+    OperationalStatusServer(
+        bus, dbusEscape(std::string(dbus::stormgr::path) + "/backplane/MCU_" +
+                        std::to_string(i2cBus) + "_" + std::to_string(i2cAddr))
+                 .c_str()),
+    i2cBusDev("/dev/i2c-" + std::to_string(i2cBus)), i2cAddr(i2cAddr),
+    cfg(config), cachedState(0)
+{
+    refresh();
+}
+
+void BackplaneController::updateConfig(const BackplaneControllerConfig& config)
+{
+    if (cfg == config)
+    {
+        return;
+    }
+    cfg = config;
+    refresh();
+}
+
+bool BackplaneController::refresh()
+{
+    const bool res = doRefresh();
+    functional(res);
+    return res;
+}
+
+bool BackplaneController::doRefresh()
+{
+    int res;
+    static const std::regex isPrintableRegex("^[[:print:]]+$",
+                                             std::regex::optimize);
+    i2cDev dev(i2cBusDev, i2cAddr);
+    if (!dev.isOk())
+    {
+        return false;
+    }
+
+    if (firmwareVersion().empty())
+    {
+        // Read MCU ident
+        res = dev.read_byte_data(OPC_GET_IDENT);
+        if (res != OPC_IDENT_RESP)
+        {
+            std::string reason;
+            if (res < 0)
+            {
+                reason = std::strerror(-res);
+            }
+            else
+            {
+                reason = "Unexpected IDENT answer";
+            }
+            log<level::ERR>("Failed to read IDENT",
+                            entry("BUS=%s", i2cBusDev.c_str()),
+                            entry("ADDR=%d", i2cAddr), entry("RESULT=%d", res),
+                            entry("REASON=%s", reason.c_str()));
+            return false;
+        }
+
+        // Read MCU version
+        unsigned char data[60];
+        memset(data, 0, sizeof(data));
+        std::string value;
```

Comment:
What's the benefit of double buffering?
In my opinion it brings negative impact during each new allocation as it uses a mutex that may lead to significant delay.

Reply:
ack